### PR TITLE
[FLINK-3118] [Gelly] Consider ResultTypeQueryable in Message and GatherFunction

### PR DESCRIPTION
**Pull request by s1ck**

### CODE REVIEW

File: flink-java/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java [hunk 1]
```diff
@@ -340,6 +340,15 @@ protected TypeExtractor() {
 		}
 		return ti;
 	}
+
+	@SuppressWarnings("unchecked")
+	public static <OUT> TypeInformation<OUT> createTypeInfo(Class<?> baseClass, Class<?> clazz, int returnParamPos, Object instance) {
+		if (instance != null && instance instanceof ResultTypeQueryable) {
```
Comment (tillrohrmann) [inline flink-java/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java]: `x instanceof T` is automatically false if `x == null`.

File: flink-java/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java [hunk 2]
```diff
@@ -340,6 +340,15 @@ protected TypeExtractor() {
 		}
 		return ti;
 	}
+
+	@SuppressWarnings("unchecked")
+	public static <OUT> TypeInformation<OUT> createTypeInfo(Class<?> baseClass, Class<?> clazz, int returnParamPos, Object instance) {
```
Comment (tillrohrmann) [inline flink-java/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java]: Missing java docs

Comment (tillrohrmann) [inline flink-java/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java]: I would maybe move the `instance` parameter to the front of the parameter list. Then it's similar to `getUnaryOperatorReturnType`.

**Comment by tillrohrmann** [comment]
Changes look good to me. Maybe we could add a test case which verifies the behaviour. Apart from that, +1 for merging :-)

**Comment by s1ck** [comment]
Sorry for missing javadoc, I was inspired by the other methods in `TypeExtractor` ;) The test will be a generic version of `LabelPropagation`, described in https://issues.apache.org/jira/browse/FLINK-3122 Would that be enough for a test?

**Comment by vasia** [comment]
Thanks a lot for the PR @s1ck!
How about adding a test case in `TypeExtractorTest`, where you create an instance that implements `ResultTypeQueryable` and one that doesn't and you check that the new method derives the type information correctly? This should be really easy to add :)

**Comment by s1ck** [comment]
Hi, sorry for the delay. I'll add the test asap.

**Comment by vasia** [comment]
Merging this!

**Pull request closed without merging**